models/choral.js

```diff
@@ -125,6 +125,89 @@ choralSchema.statics.findDevicesForUser = function (user, cb) {
   });
 };
 
+choralSchema.statics.findTreeForUser = function (user, cb) {
+  this.find({ userId: user._id }, (err, chorals) => {
+    if (err) return cb(err, null);
+
+    var all = {};
+    var children = [];
+    for (i in chorals) {
+      children = children.concat(chorals[i].children);
+      all[chorals[i]._id] = chorals[i];
+    }
+    // Find all roots
+    this.find({ $and: [ { userId: user._id }, 
+                        { _id: { $nin: children } } ] }, (err, rootChorals) => {
+      if (err) return cb(err, null);
+
+      var nodes = {};
+      var edges = {};
+      var choralIdNode = {};
+      var choralIdEdge = {};
+
+      // Construct nodes and edges for all chorals
+      for (i in chorals) {
+        var node = {
+          id: chorals[i].choralId,
+          label: chorals[i].name,
+          shape: chorals[i].choralType == 'choral' ? 'circle' : 'square'
+        }
+
+        var e = [];
+
+        nodes[chorals[i].choralId] = node;
+
+        var edge = {};
+        if (chorals[i].children.length > 0) {
+          for (j in chorals[i].children) {
+            edge.from = chorals[i].choralId;
+            var choral = all[chorals[i].children[j].toString()];
+            if (choral) {
+              edge.to = choral.choralId;
+              e.push(edge);
+              edge = {};
+            }
+          }
+        }
+
+        edges[chorals[i].choralId] = e;
+      }
+
+      // Construct root nodes and edges
+      for (i in rootChorals) {
+        var node = {
+          id: rootChorals[i].choralId,
+          label: rootChorals[i].name,
+          shape: rootChorals[i].choralType == 'choral' ? 'circle' : 'square'
+        }
+        var n = [node];
+        var e = [];
+
+        var stack = [];
+        stack.push(rootChorals[i].choralId);
+
+        while(stack.length > 0) {
+          var v = stack.pop();
+          console.log("V: " + v);
+          var adjacent = edges[v];
+          for (j in adjacent) {
+            if (adjacent[j].from == v) {
+              stack.push(adjacent[j].to);
+              e.push(adjacent[j]);
+              n.push(nodes[adjacent[j].to]);
+            }
+          }
+        }
+
+        choralIdNode[rootChorals[i].choralId] = n;
+        choralIdEdge[rootChorals[i].choralId] = e;
+      }
+
+      cb(null, chorals, rootChorals, choralIdNode, choralIdEdge);
```

Comment:
Can we return an object instead of having 5 parameters.

cb(null, {chorals, rootChorals, choralIdNode, choralIdEdge} )